### PR TITLE
[GAWB-2648] Leo call sam to create pet service account at cluster creation time

### PR DESCRIPTION
**Pull request by rtitle**
- Removed `serviceAccount` from `ClusterRequest`
- Leo now calls sam via `HttpSamDAO` at cluster creation time to obtain a pet service account for the user
- Leo also adds/removes the Dataproc Worker role to the SA, which is needed to create the cluster

This is all working for me in a FiaB. Testing notes: https://docs.google.com/document/d/1sVU3KEnJuut-_k58K2yqjAC4zlUXFAfN0Sv35cLxN34/edit?ts=59ea31d7

I'd like to ask reviewers to try and follow the steps in that Google doc to make sure everything works.

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class CallToSamFailedException(uri: Uri, status: StatusCode, msg: String)
+  extends LeoException(s"Call to Sam endpoint [${uri.path}] failed with status $status. Message: $msg", status)
+
+/**
+  * Created by rtitle on 10/16/17.
+  */
+class HttpSamDAO(val baseSamServiceURL: String)(implicit system: ActorSystem, materializer: ActorMaterializer, executionContext: ExecutionContext) extends SamDAO {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala]: I thought about putting `SamDAO/HttpSamDAO/MockSamDAO` in workbench-libs, but wasn't really sure about the right place for them. Ideally Sam would publish its own client library as a dependency. So I just left it in Leo, with the intention that this could be moved later.

Note also this isn't a complete implementation of the Sam API; just the stuff Leo needs.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  private val samServiceURL = baseSamServiceURL
+
+  val http = Http(system)
+
+  override def getStatus(): Future[SubsystemStatus] = {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala]: It occurred to me that we should implement a Leo /status endpoint soon, since it's going to production.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 3]
```diff
@@ -39,47 +41,47 @@ case class IllegalLabelKeyException(labelKey: String)
...
+            dataAccess.clusterQuery.save(augmentedCluster, GcsPath(initBucket, GcsRelativePath("")))
+          }
+        } andThen { case Success(cluster) =>
+          // Notify the cluster monitor upon cluster creation
           clusterMonitorSupervisor ! ClusterCreated(cluster)
         }
     }
   }
 
-  def processClusterRequest(googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): ClusterRequest = {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Renamed to `addClusterDefaultLabels` and made private

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 4]
```diff
@@ -39,47 +41,47 @@ case class IllegalLabelKeyException(labelKey: String)
...
-          dataAccess.clusterQuery.save(cluster, GcsPath(initBucket, GcsRelativePath("")))
-        }
-      }
+  def createCluster(userInfo: UserInfo, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
+    samDAO.getPetServiceAccount(userInfo).flatMap { serviceAccount =>
+      createCluster(serviceAccount, googleProject, clusterName, clusterRequest)
     }
+  }
 
+  def createCluster(serviceAccount: WorkbenchUserServiceAccountEmail, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: These 2 variants of `createCluster` exist to support cluster recreation-on-failure. On cluster recreation, we already have the service account stored in the DB, so we don't need to call Sam again.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala (hunk 2 above)
Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/HttpSamDAO.scala]: 👍 , though split things into separate PRs when you can (or at least separate commits) (don't worry about it this time)

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 5]
```diff
@@ -133,8 +132,10 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig, protected va
       _ <- gdDAO.updateFirewallRule(googleProject)
       // Create the bucket in leo's google bucket and populate with initialization files
       initBucketPath <- initializeBucket(dataprocConfig.leoGoogleProject, clusterName, bucketName, clusterRequest)
+      // Add Dataproc Worker role to the pet service account
+      _ <- googleIamDAO.addIamRolesForUser(petServiceAccountConfig.googleProject.string, serviceAccount, Set("roles/dataproc.worker"))
       // Once the bucket is ready, build the cluster
-      cluster <- gdDAO.createCluster(googleProject, clusterName, clusterRequest, bucketName).andThen { case Failure(e) =>
+      cluster <- gdDAO.createCluster(googleProject, clusterName, clusterRequest, bucketName, serviceAccount).andThen { case Failure(_) =>
```
[the remark this answers is not in the document]
Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: or `onFailure`

File: src/main/resources/swagger/api-docs.yaml [hunk 6]
```diff
@@ -357,15 +357,11 @@ definitions:
     description: ''
     required:
       - bucketPath
-      - serviceAccount
```
Comment (ansingh7115) [inline src/main/resources/swagger/api-docs.yaml]: heads up - this will probably break our demo ui

Comment (rtitle) [reply src/main/resources/swagger/api-docs.yaml]: Yes! I meant to bring this up -- this PR will totally break the demo we've been doing thus far, because we'll no longer be using the special leo-demo account, but instead the user permissions in dev, which probably won't work calling back to FC prod. Maybe I'll follow up with Brad on this -- is it OK to temporarily break the demo until we get to prod, or should we build some kind of backdoor where we can still use the leo-demo credentials?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 7]
```diff
@@ -133,8 +132,10 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig, protected va
       _ <- gdDAO.updateFirewallRule(googleProject)
       // Create the bucket in leo's google bucket and populate with initialization files
       initBucketPath <- initializeBucket(dataprocConfig.leoGoogleProject, clusterName, bucketName, clusterRequest)
+      // Add Dataproc Worker role to the pet service account
+      _ <- googleIamDAO.addIamRolesForUser(petServiceAccountConfig.googleProject.string, serviceAccount, Set("roles/dataproc.worker"))
```
[the remark this answers is not in the document]
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: This is modifying the pet service account, so I think it should be the project the service account lives in, NOT the project the cluster lives in.

The project the service account lives in is a global config per environment -- right now it's just set to broad-dsde-dev (see https://github.com/broadinstitute/firecloud-develop/pull/737).

This reminds me that I still need to test a multi-project scenario, where the cluster project is actually different than broad-dsde-dev. I was having trouble setting that up within a FiaB environment.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala [hunk 8]
```diff
@@ -126,14 +127,19 @@ class ClusterMonitorActor(val cluster: Cluster,
         }
     }
 
+    // Then remove the Dataproc Worker IAM role for the pet service account
+    val iamFuture = deleteBucketFuture flatMap { _ =>
+      googleIamDAO.removeIamRolesForUser(cluster.googleProject.string, cluster.googleServiceAccount, Set("roles/dataproc.worker"))
```
[the remark this answers is not in the document]
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala]: Actually since this is modifying the pet SA, this should probably be `petServiceAccountConfig.googleProject.string`. :) Unless I'm missing something

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: `recover` takes a `Throwable => U` and returns a `Future[U]` (i.e. it maps over the exception)
`onFailure` takes a `Throwable => U` and returns a `Unit` (i.e. for registering callbacks)
`andThen` takes a `Throwable => U` and returns the original `Future[T]` (i.e. it runs a side effect)

All are possible to use, but I thought `andThen` is most concise -- I want to return the result of the _original_ future, and perform a side effect on failures.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: According to the code in workbench-libs you are setting the policy on the project, which makes sense:
https://github.com/broadinstitute/workbench-libs/pull/33/files#diff-8b5ce700eb15808e064aded0caf0b9c6R87

The permission is to be able to create dataprocs _on that project_. You pass in the user email to specify which user to apply it to, which could just as well be a human user, in which case their Google identity isn't associated with a project at all.

TLDR I'm pretty sure this should be the project associated with the cluster.

[the remark this answers is not in the document]
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Oh right. Will make that change, and also work on testing this for real. :)

File: src/main/resources/swagger/api-docs.yaml (hunk 6 above)
Comment (rtitle) [reply src/main/resources/swagger/api-docs.yaml]: Brad said ok to break the demo until we go to prod

Comment (ansingh7115) [reply src/main/resources/swagger/api-docs.yaml]: Yeah - as long as we don't have to give the demo again for a while - should be fine

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala [hunk 9]
```diff
@@ -98,21 +99,21 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig, protected
...
       //Make a Leo cluster from the Google operation details
-      LeoCluster.create(clusterRequest, clusterName, googleProject, getOperationUUID(operation), OperationName(operation.getName))
+      LeoCluster.create(clusterRequest, clusterName, googleProject, getOperationUUID(operation), OperationName(operation.getName), serviceAccount)
     }
   }
 
 
   /* Kicks off building the cluster. This will return before the cluster finishes creating. */
-  private def buildCluster(googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest, bucketName: GcsBucketName)(implicit executionContext: ExecutionContext): Future[DataprocOperation] = {
+  private def buildCluster(googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest, bucketName: GcsBucketName, serviceAccount: WorkbenchUserServiceAccountEmail)(implicit executionContext: ExecutionContext): Future[DataprocOperation] = {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala]: no longer needs `clusterRequest`

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala [hunk 10]
```diff
@@ -148,8 +154,15 @@ class ClusterMonitorActor(val cluster: Cluster,
     * @return ShutdownActor
     */
   private def handleFailedCluster(errorDetails: ClusterErrorDetails): Future[ClusterMonitorMessage] = {
-    // Delete the cluster in Google
-    gdDAO.deleteCluster(cluster.googleProject, cluster.clusterName).flatMap { _ =>
+    val deleteFuture = for {
+      // Delete the cluster in Google
+      _ <- gdDAO.deleteCluster(cluster.googleProject, cluster.clusterName)
+      // Remove the Dataproc Worker IAM role for the pet service account
+      _ <- googleIamDAO.removeIamRolesForUser(cluster.googleProject.string, cluster.googleServiceAccount, Set("roles/dataproc.worker"))
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala]: Do we want this to happen even if `deleteCluster()` fails?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 4 above)
Comment (jmthibault79) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Could you add comments for the two variants?

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 11]
```diff
@@ -39,47 +41,47 @@ case class IllegalLabelKeyException(labelKey: String)
...
-class LeonardoService(protected val dataprocConfig: DataprocConfig, protected val clusterResourcesConfig: ClusterResourcesConfig, protected val proxyConfig: ProxyConfig, gdDAO: DataprocDAO, dbRef: DbReference, val clusterMonitorSupervisor: ActorRef)(implicit val executionContext: ExecutionContext) extends LazyLogging {
+class LeonardoService(protected val dataprocConfig: DataprocConfig,
+                      protected val clusterResourcesConfig: ClusterResourcesConfig,
+                      protected val proxyConfig: ProxyConfig,
+                      protected val petServiceAccountConfig: PetServiceAccountConfig,
+                      protected val gdDAO: DataprocDAO,
+                      protected val googleIamDAO: GoogleIamDAO,
+                      protected val dbRef: DbReference,
+                      protected val clusterMonitorSupervisor: ActorRef,
+                      protected val samDAO: SamDAO)
```
Comment (ansingh7115) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: This isn't really a review suggestion - more a comment: this is getting a bit crowded - unsure if we'll need to add many more parameters in the future though

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 12]
```diff
@@ -39,47 +41,47 @@ case class IllegalLabelKeyException(labelKey: String)
...
+  def createCluster(serviceAccount: WorkbenchUserServiceAccountEmail, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
     // Check if the google project has a cluster with the same name. If not, we can create it
     dbRef.inTransaction { dataAccess =>
       dataAccess.clusterQuery.getActiveClusterByName(googleProject, clusterName)
     } flatMap {
       case Some(_) => throw ClusterAlreadyExistsException(googleProject, clusterName)
       case None =>
-        create andThen { case Success(cluster) =>
+        createGoogleCluster(serviceAccount, googleProject, clusterName, clusterRequest).flatMap { case (cluster, initBucket) =>
+          val augmentedCluster = addClusterDefaultLabels(cluster)
```
Comment (ansingh7115) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: I think if I try to create a cluster with a label with a forbidden key, it will create the cluster before checking the cluster request for forbidden keys.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 13]
```diff
@@ -225,4 +226,16 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig, protected va
...
+
+  private[service] def addClusterDefaultLabels(cluster: Cluster): Cluster = {
+    // create a LabelMap of default labels
+    val defaultLabels = DefaultLabels(cluster.clusterName, cluster.googleProject, cluster.googleBucket, cluster.googleServiceAccount, cluster.jupyterExtensionUri)
+      .toJson.asJsObject.fields.mapValues(labelValue => labelValue.convertTo[String])
+    // combine default and given labels
+    val allLabels = cluster.labels ++ defaultLabels
+    // check the labels do not contain forbidden keys
+    if (allLabels.contains(includeDeletedKey))
+      throw IllegalLabelKeyException(includeDeletedKey)
```
Comment (ansingh7115) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: As it is now, this will be thrown after the dataproc cluster is already created in Google - this check should happen before that

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala (hunk 10 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala]: Yeah probably.. will fix

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 12 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Ah hm that's true. Not sure why I moved it here - I think it could go before `createGoogleCluster` as before. Good catch!

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 13 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Right, will fix

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 11 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: We could have a `LeonardoConfig` case class which contains sub-config case classes. Then we just pass around a `LeonardoConfig` instance instead of individual ones.. .just an idea.

**Comment by rtitle** [comment]
@helgridly @jmthibault79 @ansingh7115 this is ready to re-review (and test!) My testing steps are working for me in fiab-rtitle-close-collie.

File: src/test/resources/reference.conf [hunk 14]
```diff
@@ -65,6 +65,14 @@ clusterResources {
   proxySiteConf = "site.conf"
 }
 
+sam {
+  server = "https://sam.test.org:443"
+}
+
+petServiceAccount {
+  googleProject = "my-pet-project"
+}
```
Comment (helgridly) [inline src/test/resources/reference.conf]: this can go away now, right?

Comment (rtitle) [reply src/test/resources/reference.conf]: Oops! thanks

**Comment by jmthibault79** [comment]
Tested via https://docs.google.com/document/d/1sVU3KEnJuut-_k58K2yqjAC4zlUXFAfN0Sv35cLxN34/edit?ts=59f0cd4c#

**Comment by rtitle** [comment]
Thanks all, I just need Sam approval now :)
